ExternalLibs/libressl-3.6.1/man/X509_cmp.3
.\" $OpenBSD: X509_cmp.3,v 1.3 2021/07/02 10:50:39 schwarze Exp $
.\" full merge up to: OpenSSL ea5d4b89 Jun 6 11:42:02 2019 +0800
.\"
.\" This file is a derived work.
.\" The changes are covered by the following Copyright and license:
.\"
.\" Copyright (c) 2019 Ingo Schwarze <schwarze@openbsd.org>
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.\" The original file was written by Paul Yang <yang.yang@baishancloud.com>.
.\" Copyright (c) 2019 The OpenSSL Project.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\"
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\"
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in
.\"    the documentation and/or other materials provided with the
.\"    distribution.
.\"
.\" 3. All advertising materials mentioning features or use of this
.\"    software must display the following acknowledgment:
.\"    "This product includes software developed by the OpenSSL Project
.\"    for use in the OpenSSL Toolkit. (http://www.openssl.org/)"
.\"
.\" 4. The names "OpenSSL Toolkit" and "OpenSSL Project" must not be used to
.\"    endorse or promote products derived from this software without
.\"    prior written permission. For written permission, please contact
.\"    openssl-core@openssl.org.
.\"
.\" 5. Products derived from this software may not be called "OpenSSL"
.\"    nor may "OpenSSL" appear in their names without prior written
.\"    permission of the OpenSSL Project.
.\"
.\" 6. Redistributions of any form whatsoever must retain the following
.\"    acknowledgment:
.\"    "This product includes software developed by the OpenSSL Project
.\"    for use in the OpenSSL Toolkit (http://www.openssl.org/)"
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE OpenSSL PROJECT ``AS IS'' AND ANY
.\" EXPRESSED OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
.\" PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE OpenSSL PROJECT OR
.\" ITS CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
.\" SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
.\" NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
.\" LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
.\" STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
.\" ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
.\" OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd $Mdocdate: July 2 2021 $
.Dt X509_CMP 3
.Os
.Sh NAME
.Nm X509_cmp ,
.Nm X509_NAME_cmp ,
.\" The alias X509_name_cmp(3) is intentionally undocumented
.\" because it is almost unused in real-world code.
.Nm X509_issuer_and_serial_cmp ,
.Nm X509_issuer_name_cmp ,
.Nm X509_subject_name_cmp ,
.Nm X509_CRL_cmp ,
.Nm X509_CRL_match
.Nd compare X.509 certificates and related values
.\" The function name_cmp() is intentionally undocumented.
.\" It was a mistake to make it public in the first place,
.\" and it is no longer part of the public API in OpenSSL 1.1.
.Sh SYNOPSIS
.In openssl/x509.h
.Ft int
.Fo X509_cmp
.Fa "const X509 *a"
.Fa "const X509 *b"
.Fc
.Ft int
.Fo X509_NAME_cmp
.Fa "const X509_NAME *a"
.Fa "const X509_NAME *b"
.Fc
.Ft int
.Fo X509_issuer_and_serial_cmp
.Fa "const X509 *a"
.Fa "const X509 *b"
.Fc
.Ft int
.Fo X509_issuer_name_cmp
.Fa "const X509 *a"
.Fa "const X509 *b"
.Fc
.Ft int
.Fo X509_subject_name_cmp
.Fa "const X509 *a"
.Fa "const X509 *b"
.Fc
.Ft int
.Fo X509_CRL_cmp
.Fa "const X509_CRL *a"
.Fa "const X509_CRL *b"
.Fc
.Ft int
.Fo X509_CRL_match
.Fa "const X509_CRL *a"
.Fa "const X509_CRL *b"
.Fc
.Sh DESCRIPTION
.Fn X509_cmp
compares two X.509 certificates using
.Xr memcmp 3
on the SHA1 hashes of their canonical (DER) representations as generated with
.Xr X509_digest 3 .
.Pp
.Fn X509_NAME_cmp
compares two X.501
.Vt Name
objects using their canonical (DER) representations generated with
.Xr i2d_X509_NAME 3 .
.Pp
.Fn X509_issuer_and_serial_cmp
compares the
.Fa issuer
and
.Fa serialNumber
fields of two
.Vt TBSCertificate
structures, using
.Fn X509_NAME_cmp
for the
.Fa issuer
fields.
.Pp
.Fn X509_issuer_name_cmp
compares the
.Fa issuer
fields of two
.Vt TBSCertificate
structures using
.Fn X509_NAME_cmp .
.Pp
.Fn X509_subject_name_cmp
compares the
.Fa subject
fields of two
.Vt TBSCertificate
structures using
.Fn X509_NAME_cmp .
.Pp
.Fn X509_CRL_cmp
is misnamed; it only compares the
.Fa issuer
fields of two
.Vt TBSCertList
structures using
.Fn X509_NAME_cmp .
.Pp
.Fn X509_CRL_match
compares two certificate revocation lists using
.Xr memcmp 3
on the SHA1 hashes of their canonical (DER) representations as generated with
.Xr X509_CRL_digest 3 .
.Sh RETURN VALUES
All these functions return 0 to indicate a match or a non-zero value
to indicate a mismatch.
.Pp
.Fn X509_NAME_cmp ,
.Fn X509_issuer_and_serial_cmp ,
.Fn X509_issuer_name_cmp ,
.Fn X509_subject_name_cmp
and
.Fn X509_CRL_cmp
may return -2 to indicate an error.
.Sh SEE ALSO
.Xr i2d_X509_NAME 3 ,
.Xr X509_CRL_new 3 ,
.Xr X509_digest 3 ,
.Xr X509_NAME_new 3 ,
.Xr X509_new 3
.Sh STANDARDS
RFC 5280: Internet X.509 Public Key Infrastructure Certificate
and Certificate Revocation List (CRL) Profile
.Bl -dash -compact -offset indent
.It
section 4.1: Basic Certificate Fields
.It
section 5.1: CRL Fields
.El
.Sh HISTORY
.Fn X509_issuer_and_serial_cmp ,
.Fn X509_issuer_name_cmp ,
and
.Fn X509_subject_name_cmp
first appeared in SSLeay 0.5.1 and
.Fn X509_NAME_cmp
and
.Fn X509_CRL_cmp
in SSLeay 0.8.0.
These functions have been available since
.Ox 2.4 .
.Pp
.Fn X509_cmp
first appeared in OpenSSL 0.9.5 and has been available since
.Ox 2.7 .
.Pp
.Fn X509_CRL_match
first appeared in OpenSSL 1.0.0 and has been available since
.Ox 4.9 .
.Sh BUGS
For
.Fn X509_NAME_cmp ,
.Fn X509_issuer_and_serial_cmp ,
.Fn X509_issuer_name_cmp ,
.Fn X509_subject_name_cmp
and
.Fn X509_CRL_cmp ,
the return value -2 sometimes indicates a mismatch and sometimes an error.


ExternalLibs/libressl-3.6.1/crypto/x509/x509_cmp.c
/* $OpenBSD: x509_cmp.c,v 1.39 2022/02/24 22:05:06 beck Exp $ */
/* Copyright (C) 1995-1998 Eric Young (eay@cryptsoft.com)
 * All rights reserved.
 *
 * This package is an SSL implementation written
 * by Eric Young (eay@cryptsoft.com).
 * The implementation was written so as to conform with Netscapes SSL.
 *
 * This library is free for commercial and non-commercial use as long as
 * the following conditions are aheared to.  The following conditions
 * apply to all code found in this distribution, be it the RC4, RSA,
 * lhash, DES, etc., code; not just the SSL code.  The SSL documentation
 * included with this distribution is covered by the same copyright terms
 * except that the holder is Tim Hudson (tjh@cryptsoft.com).
 *
 * Copyright remains Eric Young's, and as such any Copyright notices in
 * the code are not to be removed.
 * If this package is used in a product, Eric Young should be given attribution
 * as the author of the parts of the library used.
 * This can be in the form of a textual message at program startup or
 * in documentation (online or textual) provided with the package.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *    "This product includes cryptographic software written by
 *     Eric Young (eay@cryptsoft.com)"
 *    The word 'cryptographic' can be left out if the rouines from the library
 *    being used are not cryptographic related :-).
 * 4. If you include any Windows specific code (or a derivative thereof) from
 *    the apps directory (application code) you must include an acknowledgement:
 *    "This product includes software written by Tim Hudson (tjh@cryptsoft.com)"
 *
 * THIS SOFTWARE IS PROVIDED BY ERIC YOUNG ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 * The licence and distribution terms for any publically available version or
 * derivative of this code cannot be changed.  i.e. this code cannot simply be
 * copied and put under another distribution licence
 * [including the GNU Public Licence.]
 */

#include <ctype.h>
#include <stdio.h>
#include <string.h>

#include <openssl/opensslconf.h>

#include <openssl/asn1.h>
#include <openssl/err.h>
#include <openssl/objects.h>
#include <openssl/x509.h>
#include <openssl/x509v3.h>

#include "evp_locl.h"
#include "x509_lcl.h"

int
X509_issuer_and_serial_cmp(const X509 *a, const X509 *b)
{
	int i;
	X509_CINF *ai, *bi;

	ai = a->cert_info;
	bi = b->cert_info;
	i = ASN1_INTEGER_cmp(ai->serialNumber, bi->serialNumber);
	if (i)
		return (i);
	return (X509_NAME_cmp(ai->issuer, bi->issuer));
}

#ifndef OPENSSL_NO_MD5
unsigned long
X509_issuer_and_serial_hash(X509 *a)
{
	unsigned long ret = 0;
	EVP_MD_CTX ctx;
	unsigned char md[16];
	char *f;

	EVP_MD_CTX_init(&ctx);
	f = X509_NAME_oneline(a->cert_info->issuer, NULL, 0);
	if (f == NULL)
		goto err;
	if (!EVP_DigestInit_ex(&ctx, EVP_md5(), NULL))
		goto err;
	if (!EVP_DigestUpdate(&ctx, (unsigned char *)f, strlen(f)))
		goto err;
	free(f);
	f = NULL;
	if (!EVP_DigestUpdate(&ctx,
	    (unsigned char *)a->cert_info->serialNumber->data,
	    (unsigned long)a->cert_info->serialNumber->length))
		goto err;
	if (!EVP_DigestFinal_ex(&ctx, &(md[0]), NULL))
		goto err;
	ret = (((unsigned long)md[0]) | ((unsigned long)md[1] << 8L) |
	    ((unsigned long)md[2] << 16L) | ((unsigned long)md[3] << 24L)) &
	    0xffffffffL;

err:
	EVP_MD_CTX_cleanup(&ctx);
	free(f);
	return (ret);
}
#endif

int
X509_issuer_name_cmp(const X509 *a, const X509 *b)
{
	return (X509_NAME_cmp(a->cert_info->issuer, b->cert_info->issuer));
}

int
X509_subject_name_cmp(const X509 *a, const X509 *b)
{
	return (X509_NAME_cmp(a->cert_info->subject, b->cert_info->subject));
}

int
X509_CRL_cmp(const X509_CRL *a, const X509_CRL *b)
{
	return (X509_NAME_cmp(a->crl->issuer, b->crl->issuer));
}

#ifndef OPENSSL_NO_SHA
int
X509_CRL_match(const X509_CRL *a, const X509_CRL *b)
{
	return memcmp(a->hash, b->hash, X509_CRL_HASH_LEN);
}
#endif

X509_NAME *
X509_get_issuer_name(const X509 *a)
{
	return (a->cert_info->issuer);
}

unsigned long
X509_issuer_name_hash(X509 *x)
{
	return (X509_NAME_hash(x->cert_info->issuer));
}

#ifndef OPENSSL_NO_MD5
unsigned long
X509_issuer_name_hash_old(X509 *x)
{
	return (X509_NAME_hash_old(x->cert_info->issuer));
}
#endif

X509_NAME *
X509_get_subject_name(const X509 *a)
{
	return (a->cert_info->subject);
}

ASN1_INTEGER *
X509_get_serialNumber(X509 *a)
{
	return (a->cert_info->serialNumber);
}

const ASN1_INTEGER *
X509_get0_serialNumber(const X509 *a)
{
	return (a->cert_info->serialNumber);
}

unsigned long
X509_subject_name_hash(X509 *x)
{
	return (X509_NAME_hash(x->cert_info->subject));
}

#ifndef OPENSSL_NO_MD5
unsigned long
X509_subject_name_hash_old(X509 *x)
{
	return (X509_NAME_hash_old(x->cert_info->subject));
}
#endif

#ifndef OPENSSL_NO_SHA
/* Compare two certificates: they must be identical for
 * this to work. NB: Although "cmp" operations are generally
 * prototyped to take "const" arguments (eg. for use in
 * STACKs), the way X509 handling is - these operations may
 * involve ensuring the hashes are up-to-date and ensuring
 * certain cert information is cached. So this is the point
 * where the "depth-first" constification tree has to halt
 * with an evil cast.
 */
int
X509_cmp(const X509 *a, const X509 *b)
{
	/* ensure hash is valid */
	X509_check_purpose((X509 *)a, -1, 0);
	X509_check_purpose((X509 *)b, -1, 0);

	return memcmp(a->hash, b->hash, X509_CERT_HASH_LEN);
}
#endif

int
X509_NAME_cmp(const X509_NAME *a, const X509_NAME *b)
{
	int ret;

	/* Ensure canonical encoding is present and up to date */
	if (!a->canon_enc || a->modified) {
		ret = i2d_X509_NAME((X509_NAME *)a, NULL);
		if (ret < 0)
			return -2;
	}
	if (!b->canon_enc || b->modified) {
		ret = i2d_X509_NAME((X509_NAME *)b, NULL);
		if (ret < 0)
			return -2;
	}
	ret = a->canon_enclen - b->canon_enclen;
	if (ret)
		return ret;
	return memcmp(a->canon_enc, b->canon_enc, a->canon_enclen);
}

unsigned long
X509_NAME_hash(X509_NAME *x)
{
	unsigned long ret = 0;
	unsigned char md[SHA_DIGEST_LENGTH];

	/* Make sure X509_NAME structure contains valid cached encoding */
	i2d_X509_NAME(x, NULL);
	if (!EVP_Digest(x->canon_enc, x->canon_enclen, md, NULL, EVP_sha1(),
	    NULL))
		return 0;

	ret = (((unsigned long)md[0]) | ((unsigned long)md[1] << 8L) |
	    ((unsigned long)md[2] << 16L) | ((unsigned long)md[3] << 24L)) &
	    0xffffffffL;
	return (ret);
}


#ifndef OPENSSL_NO_MD5
/* I now DER encode the name and hash it.  Since I cache the DER encoding,
 * this is reasonably efficient. */

unsigned long
X509_NAME_hash_old(X509_NAME *x)
{
	EVP_MD_CTX md_ctx;
	unsigned long ret = 0;
	unsigned char md[16];

	/* Make sure X509_NAME structure contains valid cached encoding */
	i2d_X509_NAME(x, NULL);
	EVP_MD_CTX_init(&md_ctx);
	if (EVP_DigestInit_ex(&md_ctx, EVP_md5(), NULL) &&
	    EVP_DigestUpdate(&md_ctx, x->bytes->data, x->bytes->length) &&
	    EVP_DigestFinal_ex(&md_ctx, md, NULL))
		ret = (((unsigned long)md[0]) |
		    ((unsigned long)md[1] << 8L) |
		    ((unsigned long)md[2] << 16L) |
		    ((unsigned long)md[3] << 24L)) &
		    0xffffffffL;
	EVP_MD_CTX_cleanup(&md_ctx);

	return (ret);
}
#endif

/* Search a stack of X509 for a match */
X509 *
X509_find_by_issuer_and_serial(STACK_OF(X509) *sk, X509_NAME *name,
    ASN1_INTEGER *serial)
{
	int i;
	X509_CINF cinf;
	X509 x, *x509 = NULL;

	if (!sk)
		return NULL;

	x.cert_info = &cinf;
	cinf.serialNumber = serial;
	cinf.issuer = name;

	for (i = 0; i < sk_X509_num(sk); i++) {
		x509 = sk_X509_value(sk, i);
		if (X509_issuer_and_serial_cmp(x509, &x) == 0)
			return (x509);
	}
	return (NULL);
}

X509 *
X509_find_by_subject(STACK_OF(X509) *sk, X509_NAME *name)
{
	X509 *x509;
	int i;

	for (i = 0; i < sk_X509_num(sk); i++) {
		x509 = sk_X509_value(sk, i);
		if (X509_NAME_cmp(X509_get_subject_name(x509), name) == 0)
			return (x509);
	}
	return (NULL);
}

EVP_PKEY *
X509_get_pubkey(X509 *x)
{
	if (x == NULL || x->cert_info == NULL)
		return (NULL);
	return (X509_PUBKEY_get(x->cert_info->key));
}

EVP_PKEY *
X509_get0_pubkey(const X509 *x)
{
	if (x == NULL || x->cert_info == NULL)
		return (NULL);
	return (X509_PUBKEY_get0(x->cert_info->key));
}

ASN1_BIT_STRING *
X509_get0_pubkey_bitstr(const X509 *x)
{
	if (!x)
		return NULL;
	return x->cert_info->key->public_key;
}

int
X509_check_private_key(const X509 *x, const EVP_PKEY *k)
{
	const EVP_PKEY *xk;
	int ret;

	xk = X509_get0_pubkey(x);

	if (xk)
		ret = EVP_PKEY_cmp(xk, k);
	else
		ret = -2;

	switch (ret) {
	case 1:
		break;
	case 0:
		X509error(X509_R_KEY_VALUES_MISMATCH);
		break;
	case -1:
		X509error(X509_R_KEY_TYPE_MISMATCH);
		break;
	case -2:
		X509error(X509_R_UNKNOWN_KEY_TYPE);
	}
	if (ret > 0)
		return 1;
	return 0;
}

/*
 * Not strictly speaking an "up_ref" as a STACK doesn't have a reference
 * count but it has the same effect by duping the STACK and upping the ref of
 * each X509 structure.
 */
STACK_OF(X509) *
X509_chain_up_ref(STACK_OF(X509) *chain)
{
	STACK_OF(X509) *ret;
	size_t i;

	ret = sk_X509_dup(chain);
	for (i = 0; i < sk_X509_num(ret); i++)
		X509_up_ref(sk_X509_value(ret, i));

	return ret;
}
